externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkSQLEngine.scala

```diff
@@ -83,6 +84,33 @@ case class SparkSQLEngine(spark: SparkSession) extends Serverable("SparkSQLEngin
         checker,
         Duration(shutdownTimeout, TimeUnit.MILLISECONDS))
     })
+    stopEngineExec.foreach(exec => {
+      ThreadUtils.shutdown(
+        exec,
+        Duration(60, TimeUnit.SECONDS))
+    })
+  }
+
+  def gracefulStop(): Unit = {
+    val deregistered = new AtomicBoolean(false)
+    val stopTask: Runnable = () => {
+      if (!shutdown.get) {
+        if (deregistered.compareAndSet(false, true)) {
+          info(s"Spark engine is de-registering from engine discovery space.")
+          frontendServices.flatMap(_.discoveryService).foreach(_.stop())
+        }
+        val openSessionCount = backendService.sessionManager.getOpenSessionCount
+        if (openSessionCount > 0) {
+          info(s"$openSessionCount connection(s) are active, delay shutdown")
+        } else {
+          info(s"Spark engine has no open session now, terminating.")
+          stop()
```

Comment:
seems `DeRegisterWatcher` will do this, `frontendServices.flatMap(_.discoveryService).foreach(_.stop())` is enough ?

Reply:
Yes. also remove #3774